File: 0015-3sum/0015-3sum.java
class Solution {
  public List<List<Integer>> threeSum(int[] nums) {
    List<List<Integer>> ans = new ArrayList<>();
    Arrays.sort(nums);
    int i = 0;
    int l;
    int r;
    while (i < nums.length - 2) {
      l = i + 1;
      r = nums.length - 1;
      while (l < r) {
        if (nums[i] + nums[l] + nums[r] < 0) {
          l++;
        } else if (nums[i] + nums[l] + nums[r] > 0) {
          r--;
        } else {
          List<Integer> combo = Arrays.asList(nums[i], nums[l], nums[r]);
          ans.add(combo);
          while (l < nums.length - 1 && nums[l] == nums[l + 1]) {
            l++;
          }
          if (l < nums.length - 1) {
            l++;
          }
        }
      }
      while (i < nums.length - 1 && nums[i] == nums[i + 1]) {
        i++;
      }
      if (i < nums.length - 1) {
        i++;
      }
    }
    return ans;
  }
}

File: 0015-3sum/0015-3sum.go
func threeSum(nums []int) [][]int {
  ans := make([][]int, 0)
  sort.Ints(nums)
  i, l, r := 0, 0, 0
  for i < len(nums)-2 {
    l, r = i+1, len(nums)-1
    for l < r {
      if nums[i]+nums[l]+nums[r] < 0 {
        l++
      } else if nums[i]+nums[l]+nums[r] > 0 {
        r--
      } else {
        combo := []int{nums[i], nums[l], nums[r]}
        ans = append(ans, combo)
        for l < len(nums)-1 && nums[l] == nums[l+1] {
          l++
        }
        if l < len(nums)-1 {
          l++
        }
      }
    }
    for i < len(nums)-1 && nums[i] == nums[i+1] {
      i++
    }
    if i < len(nums)-1 {
      i++
    }
  }
  
  return ans
}